src/webgpu/api/validation/render_pipeline/depth_stencil_state.spec.ts

```diff
@@ -50,13 +54,38 @@ g.test('depthCompare_optional')
     t.selectDeviceOrSkipTestCase(info.feature);
   })
   .fn(t => {
-    const { isAsync, format, depthCompare } = t.params;
+    const {
+      isAsync,
+      format,
+      depthCompare,
+      depthWriteEnabled,
+      stencilFrontDepthFailOp,
+      stencilBackDepthFailOp,
+    } = t.params;
     const info = kTextureFormatInfo[format];
     const descriptor = t.getDescriptor({
-      depthStencil: { format, depthCompare, depthWriteEnabled: false },
+      depthStencil: {
+        format,
+        depthCompare,
+        depthWriteEnabled,
+        stencilFront: { depthFailOp: stencilFrontDepthFailOp },
+        stencilBack: { depthFailOp: stencilBackDepthFailOp },
+      },
     });
 
-    t.doCreateRenderPipelineTest(isAsync, !(info.depth && depthCompare === undefined), descriptor);
+    const areDepthFailOpKeep =
+      stencilFrontDepthFailOp === 'keep' && stencilBackDepthFailOp === 'keep';
+    const success =
+      // depthCompare is optional for stencil-only formats.
+      (!info.depth && depthWriteEnabled !== true) ||
+      // depthCompare is optional for formats with a depth if it is not used for anything.
```

Comment:
I wish there was a way to make this code more readable but I didn't find it.

Reply:
Often we use a pattern like
```
let success = true;
if (...) success = false;
if (...) success = false;
```

Reply:
Most of the PR looks good but if you don't mind I'll rereview after you've done a refactor like that. It's very hard to read right now.

I think it can looks something approximately like:
```
let success = false;
if (info.depth) {
} else {
  if (depthWriteEnabled === false && areDepthFailOpKeep) success = true;
  if (depthCompare) {
    if (...) success = false;
  }
}
```

Reply:
Following your suggestion, I've uploaded another patch. Thank you @kainino0x!